Improving the efficiency of the Gaussian and (Gaussian) copula methods

LHBO opened the pull request:
In this pull request, we improve the efficiency of the Gaussian and (Gaussian) copula methods.

The efficiency has been improved by
1. Computing the conditional Gaussian distributions only once and then adding the explicand-specific parts to the precomputed quantities as outlined in #231.
2. Sampling MC samples only once from a standard normal and then transforming them to the correct conditional normal distribution.
3. Rewriting the computationally heavy parts of the code into C++ code to speed up the computations.

In this pull request, the following C++ functions replace the equivalent R functions:
- The C++ function `prepare_data_gaussian_cpp` replaces the R function `prepare_data_gaussian`.
- The C++ function `prepare_data_copula_cpp` replaces the R function `prepare_data_copula`.
- The C++ function `inv_gaussian_transform_cpp` replaces the R function `inv_gaussian_transform`.
- The C++ function `quantile_type7_cpp` replaces the R function `quantile(..., type = 7)`.

Closes #231

martinju commented:
This is great! I am most in favor v5, which I added a simplification to, using rnorm directly instead of rmvnorm. This is slightly faster. I didn't check, but I think one can then also go directly to the transpose of B so that may help even more. 
Further, since we are improving code here, it would be nice to see whether the whole double-lapply think can be done with 2 basic for loops in C++. All of that code is essentially just matrix multiplication etc which is much fast in C++. It is quite much code to transform, but ChatGPT should be super-helpful here, and you can also see the other cpp-files in shapr for how to set it up.  Note that it does not matter if the c++-code is not "efficient" using loops and so on, it will anyway be fast since it is being complied. 

Feel free to try that out if you want. Otherwise, I can give it a go next week.

LHBO commented:
> This is great! I am most in favor v5, which I added a simplification to, using rnorm directly instead of rmvnorm. This is slightly faster. I didn't check, but I think one can then also go directly to the transpose of B so that may help even more. Further, since we are improving code here, it would be nice to see whether the whole double-lapply think can be done with 2 basic for loops in C++. All of that code is essentially just matrix multiplication etc which is much fast in C++. It is quite much code to transform, but ChatGPT should be super-helpful here, and you can also see the other cpp-files in shapr for how to set it up. Note that it does not matter if the c++-code is not "efficient" using loops and so on, it will anyway be fast since it is being complied.
> 
> Feel free to try that out if you want. Otherwise, I can give it a go next week.

I will look at it next week. I agree that using C++ will be beneficial

martinju commented on src/Copula.cpp:
```diff
@@ -0,0 +1,229 @@
+
+//' Transforms new data to a standardized normal distribution
+//'
+//' @details The function uses `arma::quantile(...)` which corresponds to R's `stats::quantile(..., type = 5)`.
```
Correct this. Uses type = 7.

LHBO replied:
Removed the whole sentence as we are not longer using arma::quantile()

LHBO commented:
> @Lars: All good except the manual update mentioned above, and the ties-issue mentioned on slack. I will approve the failing tests once that is fixed, and you have confirmed that the tests differences are due to sampling error only.

- Fixed the manual update mentioned above. 
- I fixed the ties issues by using the R rank function for that problem. We do not need C++ code there, as the code takes milliseconds to run.
- I added a code example in `shapr/inst/scripts/compare_copula_in_R_and_C++.R` where I compare the Shapley values produced by the R and C++ versions. The difference tends to zero when `n_samples` tends to infinity. For `n_samples = 1000000`, the max absolute difference is in the third decimal point.
